sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1452,11 +1452,23 @@ object PushPredicateThroughJoin extends Rule[LogicalPlan] with PredicateHelper {
 }
 
 /**
- * Combines two adjacent [[Limit]] operators into one, merging the
- * expressions into one single expression.
+ * 1. Eliminate [[Limit]] operators if it's child max row <= limit.
+ * 2. Combines two adjacent [[Limit]] operators into one, merging the
+ *    expressions into one single expression.
  */
-object CombineLimits extends Rule[LogicalPlan] {
-  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+object EliminateLimits extends Rule[LogicalPlan] {
+  private def canEliminate(limitExpr: Expression, childMaxRow: Option[Long]): Boolean = {
+    limitExpr.foldable &&
+      childMaxRow.isDefined &&
+      childMaxRow.get <= limitExpr.eval().toString.toInt
```

Comment:
`limitExpr.foldable && childMaxRow.exists { _ <= limitExpr.eval().toString.toLong }`?

Reply:
Looks better. `Int` is enough since we have checked the data type during analyzer.

Reply:
I'm not a jvm expert, but jvm implicitly coerces the type into long in the comparison? If so, I think it is a bit efficient to convert it into long directly.

Reply:
You are right, It will convert to Long implicitly, update it.